src/lib/pubkey/dilithium/dilithium_common/dilithium.h

```diff
@@ -0,0 +1,148 @@
+class Dilithium_PrivateKeyInternal;
+
+/**
+ * This implementation is based on
+ * https://github.com/pq-crystals/dilithium/commit/3e9b9f1412f6c7435dbeb4e10692ea58f181ee51
+ *
+ * Note that this is _not_ compatible with the round 3 submission of the NIST competition.
+ */
+class BOTAN_PUBLIC_API(3, 0) Dilithium_PublicKey : public virtual Public_Key
+   {
+   public:
+      void initialize_from_encoding(std::vector<uint8_t> pk,
+                                    DilithiumDimension dimension, DilithiumFlavor algo, DilithiumKeyEncoding encoding);
+
+      Dilithium_PublicKey& operator=(const Dilithium_PublicKey& other) = default;
+
+      virtual ~Dilithium_PublicKey() = default;
+
+      std::string algo_name() const override;
+
+      AlgorithmIdentifier algorithm_identifier() const override;
+
+      OID get_oid() const override;
+
+      size_t key_length() const override;
+
+      size_t estimated_strength() const override;
+
+      std::vector<uint8_t> public_key_bits() const override;
+
+      bool check_key(RandomNumberGenerator&, bool) const override;
+
+      std::unique_ptr<PK_Ops::Verification>
+      create_verification_op(const std::string& params,
+                             const std::string& provider) const override;
+
+      void set_binary_encoding(DilithiumKeyEncoding encoding)
+         {
+         m_key_encoding = encoding;
+         }
+
+      DilithiumKeyEncoding binary_encoding() const
+         {
+         return m_key_encoding;
+         }
+
+      Dilithium_PublicKey(const AlgorithmIdentifier& alg_id, const std::vector<uint8_t>& pk);
+
+      Dilithium_PublicKey(const std::vector<uint8_t>& pk,
+                          DilithiumDimension dimension, DilithiumFlavor algo, DilithiumKeyEncoding encoding);
+
+   protected:
+      Dilithium_PublicKey() : m_key_encoding(DilithiumKeyEncoding::Raw)
+         {
+         }
+
+      void initialize_from_encoding(std::vector<uint8_t> pub_key, DilithiumDimension m, DilithiumKeyEncoding encoding);
+
+      friend class Dilithium_Verification_Operation;
+      friend class Dilithium_Signature_Operation;
+
+      std::shared_ptr<Dilithium_PublicKeyInternal> m_public;
+      DilithiumKeyEncoding m_key_encoding;
+   };
+
+class BOTAN_PUBLIC_API(3, 0) Dilithium_PrivateKey final : public virtual Dilithium_PublicKey,
+   public virtual Botan::Private_Key
+   {
+   public:
+      std::unique_ptr<Public_Key> public_key() const override;
+
+      Dilithium_PrivateKey(RandomNumberGenerator& rng, DilithiumDimension dimension, DilithiumFlavor algo);
+
+      /**
+       * Read an encoded private key. Note that the resulting object will throw
+       * when trying to use it as a public key.
+       */
+      Dilithium_PrivateKey(const AlgorithmIdentifier& alg_id, const secure_vector<uint8_t>& sk);
+
+      /**
+       * Read an encoded private key. Polymorphically using the resulting object
+       * as a public key will throw. It that is required, use the respective
+       * constructor.
```

Comment:
This is a pretty unfortunate situation that will certainly cause confusing behavior for applications. Is there any way we can avoid this?

(`Private_Key` inheriting from `Public_Key` was absolutely a mistake, but I'm not sure if we can fix it now ...)

Reply:
The only thing I could think of would be to enforce that private keys must be instantiated with their associated public key. But that's also unfavorable.

Off the top of my head: I'm assuming that the math behind dilithium does not allow to derive the public key from the private key. To my mind, the possibility to do that for "classical" algorithms was likely the reasoning for this inheritance decision anyway, right?

Reply:
Ah yeah perhaps there is some seed material that is thrown away after the public key is generated, and is not stored in the private key? I don't remember the details - it's been a long time since I looked at the Dilithium spec and it almost certainly was before the last round of changes.

> To my mind, the possibility to do that for "classical" algorithms was likely the reasoning for this inheritance decision anyway, right?

Yes exactly. However PQC or no, it's pretty weird that you would have a public key algorithm where the public key cannot be derived from the private key.

I'll look at the spec this weekend to figure out what's going on here.

TBH though - here we are following a internet draft that is (IIRC) abandoned. Perhaps we should instead (given that we are anyway not, afaik, compatible with any other encoding, using private allocation OIDs, etc) forge our own path wrt the DER encoding of the private key, include whatever bits we need so that the library invariants are maintained, and then when the IETF gets around to standardizing Dilithium for real we/I can argue with the WG about why our method is better than this random draft ;)

Reply:
OK as I understand it the issue is that the public key is (notation from the Dilithium 3.1 spec pdf, page 13)

`(rho,t1)`
and the private key is
`(rho,K,tr,s1,s2,t0)`

we can recompute `t1` from `rho`, `s1`, s2` by

```
A = ExpandA(rho)
t = A*s1 + s2
(t1,t0) = Power2Round(t, d)
```

so we do not need the public key to be provided - we can quite cheaply recompute it. I wish they had just included `t1` in the private key as well, but this does save a few hundred bytes so I can understand the motivation.

Reply:
Indeed the code for that is already there (in the key generation). A minor refactoring allowed us to re-use it for said use case. That's a major UX improvement. Thanks for looking into this.